### ISSUE TITLE
Restarting workflow after updating local file

### ISSUE THREAD

**Issue by wpoehlm** [DataBiosphere/toil#2773]
Hi,

I was wondering if someone could clarify or give advice on a situation that we have run into.  We are running a CWL workflow using a toil appliance (toil:3.21.0a1) that uses a S3 bucket jobstore.  When we first submitted the workflow, we forgot to place files into a local directory that are necessary for one of the downstream subworkflows (the local paths to these files were defined in the input json/yaml file).  As a result, the workflow failed when it reached the subworkflow that requires these files:

`IOError: [Errno 2] No such file or directory: u'/root/jobs/Run_Rosmap_8_26_19/amp-rnaseq_reprocessing/amp-rnaseq_reprocess-workflow/Ref/gencode.v31.annotation.gtf'
ERROR:cwltool:Got workflow error`

Since the error message provided the local file path, we were hoping that restarting the workflow after adding the correct files locally would fix the issue.  However, this was not the case, which led us to believe that all expected input files are staged when the workflow is submitted, not when the workflow reaches the point that they are needed.  

Does anyone have any advice on restarting the workflow such that it finds the local files that were originally missing?  We would like to avoid restarting the workflow from the beginning since that would waste a lot of compute time.  We are testing the possibility of updating the jobstore with the necessary files, but it would be great if someone could lead us down the correct path.  I would be happy to provide more details about the workflow as needed.  

Thanks,
Will

**Comment by DailyDreaming** [comment]
@wpoehlm This sounds like a bug, as passing the `--restart` flag _should_ pick up on the new files.

@adamnovak Have you had experience with files that were ignored upon a restart?

**Comment by adamnovak** [comment]
No, but I'm not sure that the CWL runner doesn't pull files into the file store or otherwise cache their non-existence, if you deliver them through the normal CWL channels for moving files around.

**Comment by ionox0** [comment]
I've been able to successfully change the contents of an input file, keeping the same input path, and restart a workflow using `toil-cwl-runner`.

However I haven't tried with a file that was originally missing altogether.

**Comment by tthyer** [comment]
Apologies for the late response. @wpoehlm and I are working on this together, and have created a small repro case. 

Files in the attached
[repro.zip](https://github.com/DataBiosphere/toil/files/3688336/repro.zip)
 include workflow.cwl, workflow-inputs.yaml, and names.txt. This is set up to fail as the filename in workflow-inputs.yaml is "name.txt", not "names.txt". 

Expected behavior: first workflow run fails, but second workflow run (with --restart) succeeds after renaming "names.txt" to "name.txt"

Actual behavior: when running with an S3 jobstore, the second workflow run fails also. (Although, I see the expected behavior when running with a local jobstore).

Note: the Toil image Will mentioned, 3.21.0a1, is because I built our own Toil appliance in order to get the fix in this PR: https://github.com/DataBiosphere/toil/pull/2690.

**Comment by ionox0** [comment]
sorry that this doesn't solve your issue, but I don't think it's possible to correct input file paths and restart a workflow at the moment.

**Comment by tthyer** [comment]
@ionox0 I agree, it's not possible at the moment, and I suggest it is a bug. It has two parts:
1. Either toil or cwltool could verify File and Directory inputs exist with os.path. 
2. On the toil side, I suspect the problem is in cwltoil.py. At [line 1249](https://github.com/DataBiosphere/toil/blob/master/src/toil/cwl/cwltoil.py#L1249) in that file as it stands in master at the moment, it handles the restart option. If it's restarting, it just calls `toil.restart()`. But if not, it looks like this is where a lot of job setup happens, and a bit later around line 1300 on is where a file importing to the jobstore happens, I think.

**Comment by drkennetz** [comment]
Question related to this as I wasn't completely able to tell from the convo above...

If I add a parameter to a tool in cwl and try to restart a workflow from toil at the current step will it work? Or should I just recreate the run?

**Comment by ionox0** [comment]
@drkennetz It's not possible (in my experience) to change the contents of any `.cwl` workflow or tool and have those changes reflected in the new run. The workflow is built (serialized?) by cwltool right off the bat and it's contents are saved for the re-run (someone would have to correct me if I'm wrong though)

@tthyer Thanks for the detailed test case, and yes your comment about `toil.restart()` is certainly a possibility. I'm not a core dev but I'd be able to review and test a PR

**Comment by drkennetz** [comment]
@ionox0 no that is what it seems like it happening. I tried a restart after changing the CLT and didn't see it reflected in the job store. I think those files are pickled at the start, so I think theoretically it would be possible to unpickle, compare, and decide.
